### PR TITLE
Tree: Add 'Ordered-List' DataObject

### PR DESCRIPTION
**Pull request by DLehenbauer**
Example for a customer demonstrating how to use SharedTree to model an ordered list of items with intention-preserving move operations.

### CODE REVIEW

File: experimental/examples/ordered-list/src/orderedlist.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+} from "@fluid-experimental/tree";
+import { v4 as uuid } from "uuid";
+import { Serializable } from "@fluidframework/datastore-definitions";
+import { IFluidHandle } from "@fluidframework/core-interfaces";
+
+const listItemDefinition = undefined as unknown as Definition;
+const itemsTrait = "" as TraitLabel;
+
+const makePositionNode = <T>(value: Serializable<T>): ChangeNode => ({
+    identifier: uuid() as NodeId,
```
Comment (dcakirkaya) [inline experimental/examples/ordered-list/src/orderedlist.ts:28]: how does it ensure everybody ends up with the  same list when there are concurrent insertions to the same position?

File: experimental/examples/ordered-list/src/orderedlist.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        this.tree.applyEdit(
+            Delete.create(
+                StableRange.all({
+                    parent: this.view.root,
+                    label: itemsTrait,
+                })));
+    }
+
+    /** Returns an array containing the positions of all items in the list. */
+    public get positions() {
```
Comment (dcakirkaya) [inline experimental/examples/ordered-list/src/orderedlist.ts:156]: Curious, why OrderedList doesn't provide any iterator over elements but an array of positions ?
Also, if this is the way to get elements: 

```ts
list.positions.map((position) => list.get(position)); 
```
- Are position and _item at position_ guaranteed to remain the same when we do this?
- Does it return snapshot of positions  ?  (positions might have been added or removed when we call getPosition but will not be included in results?)

File: experimental/examples/ordered-list/src/orderedlist.ts (hunk 1 above)
Comment (DLehenbauer) [reply experimental/examples/ordered-list/src/orderedlist.ts:28]: The Fluid service assigns an increasing seq# to each op as it is received before broadcasting them to clients. This seq# defines a total ordering that Fluid DDSes must honor when constructing the client's local copy of the data.

The only "unordered" ops that a DDS must deal with are your client's local edits that have not yet been sequenced by the Fluid service. However, the DDS knows that all of these unordered local edits will be assigned higher sequence numbers than the remote ops it receives.

Therefore, all copies of a Fluid DDS have the same relative ordering for the ops the DDS is aware of, and consequently two DDSes that are aware of the same ops will produce identical state.

File: experimental/examples/ordered-list/src/orderedlist.ts (hunk 2 above)
Comment (DLehenbauer) [reply experimental/examples/ordered-list/src/orderedlist.ts:156]: That is the correct way to get the array of items.

Fluid schedules processes batches of ops in microtasks that run interleaved with your code, but never at the same time as your code (i.e., there's no shared memory or parallel threads.)

Therefore, the DDS will only change while you're awaiting a Promise or between asynchronous callbacks/events, but never during a synchronous block of code.  Therefore, the above code snippet is safe.

There's no reason OrderedList couldn't build the list of items and support an iterator, but there's also no advantage to having the OrderedList do this vs. you building what MFS needs on top.  (e.g., I suspect you'll want to keep items associated with their positions...  like '{ position: Position, item: T }[]').

File: experimental/examples/ordered-list/src/orderedlist.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+});
+
+export type Position = NodeId;
+
+const treeKey = "" as const;
+
+/**
+ * Example DataObject demonstrating how to model an ordered list using the SharedTree DDS.
+ */
+export class OrderedList<T> extends DataObject {
```
Comment (skylerjokiel) [inline experimental/examples/ordered-list/src/orderedlist.ts:41]: Why does the order list need to have strongly typed values <T>?

**Review by skylerjokiel — COMMENTED** [review]
Just some general API thoughts. I see this is more an example of how to do something so I'm not sure the intent here so feel free to ignore any/all of my comments. One thing I really like is how this showcases using an existing DDS to build a DataObject with a specialized API. This is a powerful pattern that I would love to keep promoting in the repo.

What I would expect to see is something akin to the C# List API. But, because it uses nodes instead of an index the API feel bloated and needs to tackle very specific requests. IMO it feel pretty unnatural.

**Comment by dcakirkaya** [comment]
Daniel it would be nice if you could show item inserted, moved, removed events in this example. Curious if there are any Fluid specific patterns to follow when wiring up & translating the events from the underlying DDSs in a DataObject.

**Comment by dcakirkaya** [comment]
Would be nice to have the DataObject raise some events (inserted, moved, removed etc) in this example. 
Curious if there are any Fluid specific patterns to follow.

**Comment by dcakirkaya** [comment]
Would be nice to have the DataObject raise some events (inserted, moved, removed etc) in this example.
Curious if there are any Fluid specific patterns to follow, also how to translate the SharedTree event (`committedEdit` and editId ?) .

File: experimental/examples/ordered-list/src/orderedlist.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+    private placeAfter(position: Position) {
+        return StablePlace.after(this.view.getSnapshotNode(position).identifier);
+    }
+
+    private move(destination: StablePlace, source: StableRange) {
+        this.tree.applyEdit(...Move.create(source, destination));
+    }
+
+    private insert(place: StablePlace, item: Serializable<T>): Position {
```
[the remark this answers is not in the document]
Comment (DLehenbauer) [reply experimental/examples/ordered-list/src/orderedlist.ts:84]: Sort of, we removed the problematic types and replaced them with the ones that work (see #6316).

**Comment by DLehenbauer** [comment]
> because it uses nodes instead of an index

@skylerjokiel - The customer needs specific semantics around concurrent moves that are only offered by the 'id' based ops used by the SharedTree implementation.  I could convert indices -> positions on the fly, but it's probably harder for this customer, who will want a durable way to refer to a position that doesn't shift while an asynchronous operation like drag/drop is in progress.  (Also, the conversion is O(n) with this tree DDS.  Post convergence, we'll deal equally well with both positions and indices.)

**Pull request closed without merging**